File: Scala/src/main/scala/TextbookTrack/Chapter07/BA7C.scala
package TextbookTrack.Chapter07

object BA7C {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Queue => MutableQueue, Map => MutableMap}

  private type DistanceMatrix = Vector[Vector[Int]]

  final case class WeightedEdge(nodeFrom: Int, nodeTo: Int, weight: Int)
  final case class AttachmentPoint(startLeaf: Int, endLeaf: Int, distanceFromStart: Int)

  class Tree(val adjacencyList: Map[Int, List[WeightedEdge]], val nrLeaves: Int) {
    import Tree.resolvePathBetweenLeaves

    private def largestLeaf: Int = adjacencyList.keysIterator.filter(_ < nrLeaves).max

    private def largestInnerNode: Int = {
      val innerNodes: Iterator[Int] = adjacencyList.keysIterator.filter(_ >= nrLeaves)
      if (innerNodes.isEmpty) nrLeaves - 1
      else innerNodes.max
    }

    def edges: Iterator[String] = {
      val orderedNodes: List[Int] = adjacencyList.keysIterator.toList.sorted
      for {
        node <- orderedNodes.iterator
        neighbours: List[WeightedEdge] = adjacencyList(node)
        WeightedEdge(_, neighbour, weight) <- neighbours
      } yield s"$node->$neighbour:$weight"
    }

    def insertLeafIntoTree(limblength: Int, leafStart: Int, leafEnd: Int, insertionDistance: Int): Tree = {
      @tailrec
      def loop(path: List[WeightedEdge], pathLength: Int): Tree = path match {
        case Nil => throw new Exception("Insertion distance should be less than the total length of the path.")
        case WeightedEdge(currentNode, nextNode, weight) :: rest =>
          val currrentPathLength: Int = pathLength + weight
          if (insertionDistance < currrentPathLength) {
            val distanceFromCurrentNode: Int = weight - (currrentPathLength - insertionDistance)
            insertLeaf(currentNode, nextNode, weight, distanceFromCurrentNode, limblength)
          } else loop(rest, currrentPathLength)
      }

      val path: List[WeightedEdge] = breadthFirstSearch(leafStart, leafEnd)
      loop(path, 0)
    }

    private def breadthFirstSearch(startNode: Int, endNode: Int): List[WeightedEdge] = {
      val distances: MutableMap[Int, Int] = MutableMap(startNode -> 0)
      val backtrack: MutableMap[Int, WeightedEdge] = MutableMap()
      val queue: MutableQueue[Int] = MutableQueue(startNode)
      while (queue.nonEmpty) {
        val node: Int = queue.dequeue()
        val neighbours: List[WeightedEdge] = adjacencyList.getOrElse(node, Nil)
        neighbours.foreach{
          case edge @ WeightedEdge(_, neighbour, weight) =>
            if (!distances.contains(neighbour)) {
              queue.enqueue(neighbour)
              distances(neighbour) = distances(node) + weight
              backtrack(neighbour) = edge
            }
        }
      }
      resolvePathBetweenLeaves(backtrack.toMap, startNode, endNode)
    }

    private def insertLeaf(currentNode: Int,
                           nextNode: Int,
                           weight: Int,
                           distanceFromCurrentNode: Int,
                           limblength: Int): Tree = {
      val nextLeaf: Int = largestLeaf + 1
      val updatedAdjacencyList: Map[Int, List[WeightedEdge]] = {
        if (distanceFromCurrentNode == 0) {
          val updatedNeighbours: List[WeightedEdge] =
            WeightedEdge(currentNode, nextLeaf, limblength) :: adjacencyList(currentNode)
          adjacencyList.updated(currentNode, updatedNeighbours) +
            (nextLeaf -> List(WeightedEdge(nextLeaf, currentNode, limblength)))
        }
        else {
          val middleNode: Int = largestInnerNode + 1
          val insertedMiddleNode1: Map[Int, List[WeightedEdge]] =
            addMiddleNode(adjacencyList, currentNode, nextNode, middleNode, distanceFromCurrentNode)
          val insertedMiddleNode2: Map[Int, List[WeightedEdge]] =
            addMiddleNode(insertedMiddleNode1, nextNode, currentNode, middleNode, weight - distanceFromCurrentNode)
          Map(
            middleNode -> List(
              WeightedEdge(middleNode, currentNode, distanceFromCurrentNode),
              WeightedEdge(middleNode, nextNode, weight - distanceFromCurrentNode),
              WeightedEdge(middleNode, nextLeaf, limblength)),
            nextLeaf -> List(WeightedEdge(nextLeaf, middleNode, limblength))
          ) ++ insertedMiddleNode2
        }
      }

      new Tree(updatedAdjacencyList, nrLeaves)
    }

    private def addMiddleNode(adjacencyList: Map[Int, List[WeightedEdge]],
                              currentNode: Int,
                              neighbour: Int,
                              middleNode: Int,
                              distance: Int): Map[Int, List[WeightedEdge]] = {
      val newNeighbours: List[WeightedEdge] =
        WeightedEdge(currentNode, middleNode, distance) :: adjacencyList(currentNode).filter(_.nodeTo != neighbour)
      adjacencyList.updated(currentNode, newNeighbours)
    }
  }

  object Tree {
    private def resolvePathBetweenLeaves(backtrack: Map[Int, WeightedEdge],
                                         startNode: Int,
                                         endNode: Int): List[WeightedEdge] = {
      @tailrec
      def loop(path: List[WeightedEdge], currentNode: Int): List[WeightedEdge] = {
        if (currentNode == startNode) path
        else {
          val edge @ WeightedEdge(previousNode, _, _) = backtrack(currentNode)
          loop(edge :: path, previousNode)
        }
      }
      loop(Nil, endNode)
    }
  }

  private def convertToIntVector(line: String): Vector[Int] = line.split(" ").map(_.toInt).toVector

  private def readDistanceMatrix(reader: Iterator[String], nrLeaves: Int): DistanceMatrix =
    reader.take(nrLeaves).map(convertToIntVector).toVector

  private def calcLimbLength(distances: DistanceMatrix, leafJ: Int, nrLeaves: Int): Int = {
    val otherLeaves: List[Int] = (0 until nrLeaves).filterNot(_ == leafJ).toList
    otherLeaves
      .iterator
      .flatMap(leafI => otherLeaves.map(leafK => (leafI, leafK)))
      .foldLeft(Int.MaxValue) {
        case (limbLength, (leafI, leafK)) =>
          val candidate: Int = (distances(leafI)(leafJ) + distances(leafJ)(leafK) - distances(leafI)(leafK)) / 2
          math.min(candidate, limbLength)
      }
  }

  private def subtractLimbLength(distances: DistanceMatrix, leaf: Int, limblength: Int): DistanceMatrix =
    distances.zipWithIndex.map {
      case (row, ix) =>
        if (ix == leaf) {
          row.zipWithIndex.map { case (item, jy) => if (jy == leaf) 0 else item - limblength }
        } else row.zipWithIndex.map{ case (item, jy) => if (jy == leaf) item - limblength else item }
    }

  private def findPathWhereLeafShouldBeAttached(distances: DistanceMatrix, leaf: Int): (Int, Int) = {
    val size: Int = distances.length
    (0 until size)
      .iterator
      .flatMap(leafI => (0 until size).map(leafK => (leafI, leafK)))
      .find{
        case (leafI, leafK) =>
          distances(leafI)(leafK) == distances(leafI)(leaf) + distances(leaf)(leafK)
      } match {
      case None => throw new Exception("No path endpoints found where the leaf should be attached.")
      case Some(leafPair) => leafPair
    }
  }

  private def removeLastRowAndColumn(baldDistances: DistanceMatrix): DistanceMatrix =
    baldDistances.map(_.dropRight(1)).dropRight(1)

  private def createInitialTree(distance: Int, nrLeaves: Int): Tree = {
    val adjacencyList: Map[Int, List[WeightedEdge]] =
      Map(0 -> List(WeightedEdge(0, 1, distance)), 1 -> List(WeightedEdge(1, 0, distance)))
    new Tree(adjacencyList, nrLeaves)
  }

  private def iterativelyReduceDistanceMatrix(distances: DistanceMatrix,
                                              nrLeaves: Int): (Tree, Map[Int, Int], Map[Int, AttachmentPoint]) = {
    @tailrec
    def loop(limblengths: Map[Int, Int],
             attachmentPoints: Map[Int, AttachmentPoint],
             distances: DistanceMatrix,
             leaf: Int): (Tree, Map[Int, Int], Map[Int, AttachmentPoint]) = {
      if (leaf <= 1) {
        val leafDistance: Int = distances(0)(1)
        val tree: Tree = createInitialTree(leafDistance, nrLeaves)
        (tree, limblengths, attachmentPoints)
      } else {
        val limblength: Int = calcLimbLength(distances, leaf, leaf + 1)
        val baldDistances: DistanceMatrix = subtractLimbLength(distances, leaf, limblength)
        val (leafI, leafK): (Int, Int) = findPathWhereLeafShouldBeAttached(baldDistances, leaf)
        val dist: Int = baldDistances(leafI)(leaf)
        val updatedLimbLengths: Map[Int, Int] = limblengths + (leaf -> limblength)
        val updatedAttachmentPoints: Map[Int, AttachmentPoint] =
          attachmentPoints + (leaf -> AttachmentPoint(leafI, leafK, dist))
        loop(updatedLimbLengths, updatedAttachmentPoints, removeLastRowAndColumn(baldDistances), leaf - 1)
      }
    }

    loop(Map(), Map(), distances, nrLeaves - 1)
  }

  def additivePhylogeny(distances: DistanceMatrix, nrLeaves: Int): Tree = {
    val (initialTree, limbLengths, attachmentPoints): (Tree, Map[Int, Int], Map[Int, AttachmentPoint]) =
      iterativelyReduceDistanceMatrix(distances, nrLeaves)
      (2 until nrLeaves).foldLeft(initialTree){
        case (tree, k) =>
          val limblength: Int = limbLengths(k)
          val AttachmentPoint(leafI, leafK, distance) = attachmentPoints(k)
          tree.insertLeafIntoTree(limblength, leafI, leafK, distance)
      }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrLeaves: Int = reader.next().toInt
    val distances: DistanceMatrix = readDistanceMatrix(reader, nrLeaves)
    val result: Tree = additivePhylogeny(distances, nrLeaves)
    result.edges.foreach(println)
  }
}

File: Python/textbook_track/chapter07/ba7c.py
# Implement AdditivePhylogeny
import sys

from collections import deque
from itertools import count
from math import inf


class Tree:
    def __init__(self, adjacency_list, nr_leaves, nr_initial_leaves=2):
        self._adjacency_list = adjacency_list
        self._inner_node_generator = count(nr_leaves)
        self._leaf_generator = count(nr_initial_leaves)

    def edges(self):
        ordered_nodes = sorted(self._adjacency_list.keys())
        for node in ordered_nodes:
            neighbours = self._adjacency_list[node]
            for neighbour, weight in neighbours:
                yield f'{node}->{neighbour}:{weight}'

    def insert_leaf_into_tree(self, limblength, leaf_start, leaf_end, insertion_distance):
        path = self._breadth_first_search(leaf_start, leaf_end)
        current_node = leaf_start
        path_length = 0
        for next_node, weight in path:
            path_length += weight
            if insertion_distance < path_length:
                distance_from_current_node = weight - (path_length - insertion_distance)
                return self._insert_leaf(current_node, next_node, weight, distance_from_current_node, limblength)
            current_node = next_node

    def _breadth_first_search(self, start_leaf, end_leaf):
        distances = {start_leaf: 0}
        backtrack = {}
        queue = deque([start_leaf])
        while queue:
            node = queue.popleft()
            neighbours = self._adjacency_list.get(node, [])
            for neighbour, weight in neighbours:
                if neighbour not in distances:
                    queue.append(neighbour)
                    distances[neighbour] = distances[node] + weight
                    backtrack[neighbour] = (node, weight)
        return self._resolve_path_between_leaves(backtrack, start_leaf, end_leaf)

    @staticmethod
    def _resolve_path_between_leaves(backtrack, start_leaf, end_leaf):
        path = []
        node = end_leaf
        while node != start_leaf:
            previous_node, weight = backtrack[node]
            path.append((node, weight))
            node = previous_node
        return path[::-1]

    def _insert_leaf(self, current_node, next_node, weight, distance_from_current_node, limblength):
        next_leaf = next(self._leaf_generator)
        if distance_from_current_node == 0:
            self._adjacency_list[next_leaf] = [(current_node, limblength)]
            self._adjacency_list[current_node].append((next_leaf, limblength))
        else:
            middle_node = next(self._inner_node_generator)
            self._add_middle_node(current_node, next_node, middle_node, distance_from_current_node)
            self._add_middle_node(next_node, current_node, middle_node, weight - distance_from_current_node)
            self._adjacency_list[middle_node] = [(current_node, distance_from_current_node),
                                                 (next_node, weight - distance_from_current_node),
                                                 (next_leaf, limblength)]
            self._adjacency_list[next_leaf] = [(middle_node, limblength)]

    def _add_middle_node(self, current_node, neighbour, middle_node, distance):
        all_neighbours = self._adjacency_list[current_node]
        new_neighbours = [(node, weight) for node, weight in all_neighbours if node != neighbour]
        new_neighbours.append((middle_node, distance))
        self._adjacency_list[current_node] = new_neighbours


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_distance_matrix(reader, nr_leaves):
    matrix = []
    for _ in range(nr_leaves):
        row = convert_to_intlist(next(reader))
        matrix.append(row)
    return matrix


def calc_limb_length(distances, leaf_j, nr_leaves):
    limb_length = inf
    other_leaves = [leaf for leaf in range(nr_leaves) if leaf != leaf_j]
    for leaf_i in other_leaves:
        for leaf_k in other_leaves:
            candidate = (distances[leaf_i][leaf_j] + distances[leaf_j][leaf_k] - distances[leaf_i][leaf_k]) // 2
            if candidate < limb_length:
                limb_length = candidate
    return limb_length


def subtract_limb_length(distances, leaf, limblength):
    bald_distances = []
    for ix, row in enumerate(distances):
        updated_row = []
        for jy, elem in enumerate(row):
            item = elem
            if ix == leaf or jy == leaf:
                item -= limblength
            updated_row.append(item)
        bald_distances.append(updated_row)
    bald_distances[leaf][leaf] = 0
    return bald_distances


def find_path_where_leaf_should_be_attached(distances, leaf):
    size = len(distances)
    for leaf_i in range(size):
        for leaf_k in range(size):
            if distances[leaf_i][leaf_k] == distances[leaf_i][leaf] + distances[leaf][leaf_k]:
                return leaf_i, leaf_k


def remove_last_row_and_column(distances):
    _ = distances.pop(-1)
    for row in distances:
        _ = row.pop(-1)


def create_initial_tree(distance, nr_leaves):
    adjacency_list = {0: [(1, distance)], 1: [(0, distance)]}
    return Tree(adjacency_list, nr_leaves)


def iteratively_reduce_distance_matrix(distances, nr_leaves):
    limblengths = [None] * nr_leaves
    attachment_points = [None] * nr_leaves
    for last_leaf in range(nr_leaves-1, 1, -1):
        limblength = calc_limb_length(distances, last_leaf, last_leaf+1)
        limblengths[last_leaf] = limblength
        bald_distances = subtract_limb_length(distances, last_leaf, limblength)
        leaf_i, leaf_k = find_path_where_leaf_should_be_attached(bald_distances, last_leaf)
        dist = bald_distances[leaf_i][last_leaf]
        attachment_points[last_leaf] = (leaf_i, leaf_k, dist)
        remove_last_row_and_column(bald_distances)
        distances = bald_distances
    initial_tree = create_initial_tree(distances[0][1], nr_leaves)
    return initial_tree, limblengths, attachment_points


def additive_phylogeny(distances, nr_leaves):
    tree, limblengths, attachment_points = iteratively_reduce_distance_matrix(distances, nr_leaves)
    for limblength, (leaf_i, leaf_k, distance) in zip(limblengths[2:], attachment_points[2:]):
        tree.insert_leaf_into_tree(limblength, leaf_i, leaf_k, distance)
    return tree


def main():
    reader = sys.stdin
    nr_leaves = int(next(reader))
    distances = read_distance_matrix(reader, nr_leaves)
    result = additive_phylogeny(distances, nr_leaves)
    for edge in result.edges():
        print(edge)


if __name__ == '__main__':
    main()
